modules/concurrency/src/spinner.cpp

```diff
@@ -64,34 +81,60 @@ void Spinner::spin() {
 
   try {
     start_timestamp_ = std::chrono::system_clock::now();
+    uint64_t spin_count_{ 0 };
 
     while (!stop_requested_.load()) {
-      const auto spin_result = stoppable_callback_();
-
-      ++spin_count_;
-
-      if (spin_result == SpinResult::STOP) {
+      state_ = attemptBinaryTransition(state_, { .input_state = State::NOT_INITIALIZED,
+                                                 .operation = callbacks_.init_cb,
+                                                 .success_state = State::INIT_SUCCESSFUL,
+                                                 .failure_state = State::INIT_FAILED });
+
+      state_ = attemptBinaryTransition(state_, { .input_state = State::INIT_FAILED,
+                                                 .operation = callbacks_.shall_re_init_cb,
+                                                 .success_state = State::NOT_INITIALIZED,
+                                                 .failure_state = State::TERMINATE });
+
+      state_ = attemptBinaryTransition(state_, { .input_state = State::READY_TO_SPIN,
+                                                 .operation = callbacks_.spin_once_cb,
+                                                 .success_state = State::SPIN_SUCCESSFUL,
+                                                 .failure_state = State::SPIN_FAILED });
+
+      state_ = attemptBinaryTransition(state_, { .input_state = State::SPIN_FAILED,
+                                                 .operation = callbacks_.shall_re_init_cb,
+                                                 .success_state = State::NOT_INITIALIZED,
+                                                 .failure_state = State::TERMINATE });
```

Comment:
I think `shall_re_init_cb` should be associated only to init failures, here we should call, `shall_stop`